rep-2011.rst

```diff
@@ -0,0 +1,1421 @@
+
+    string type_name
+    Field[] fields
+
+And the ``Field`` type:
+
+.. code::
+
+    FIELD_TYPE_NESTED_TYPE = 0
+    FIELD_TYPE_INT = 1
+    FIELD_TYPE_DOUBLE = 2
+    # ... and so on
+    FIELD_TYPE_INT_ARRAY = ...
+    FIELD_TYPE_INT_BOUNDED_SEQUENCE = ...
+    FIELD_TYPE_INT_SEQUENCE = ...
+    # ... and so on
+
+    string field_name
+    uint8_t field_type
+    uint64_t field_array_size  # Only for Arrays and Bounded Sequences
+    string nested_type_name  # Only for FIELD_TYPE_NESTED_TYPE
+
+These examples of the interfaces just give an idea of the structure but perhaps do not yet consider some other complications like field annotations or other as yet unconsidered features we want to support.
+
+.. TODO::
+
+    (wjwwood) Add text about how to handle Service types, which are formed as a Request and a Response part, each of which is kind of like a Message.
+    We could just treat the Request and Response separately, or we could extend this scheme to include explicit support for Services.
+    Since Actions are composed of Topics and Services, it is less so impacted, but we could similarly consider officially supporting them in this scheme.
+
+Versioning the ``TypeDescription`` Message Type
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Given that the type description message interface has to be generic enough to support anything described in the ROS interfaces, there will be a need to add or remove fields over time in the type description message itself.
+This should be done in such a way that the fields are tick-tocked and deprecated properly, possibly by having explicitly named versions of this interface, e.g. ``TypeDescriptionV1`` and ``TypeDescriptionV2`` and so on.
+
+Implementation in the ``rcl`` Layer
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The implementation of the type description distribution feature will be made in the ``rcl`` layer as opposed to the ``rmw`` layer to take advantage of the abstraction away from the middleware and to allow for compatibility with the client libraries.
+
+A hook will be added to ``rcl_node_init()`` to initialize the type description distribution service with the appropriate ``rcl_service_XXX()`` functions.
+This hook should also keep a map of published and subscribed types which will be populated on each initialization of a publisher or subscription in the respective ``rcl_publisher_init()`` and ``rcl_subscription_init()`` function calls.
+The passed ``rosidl_message_type_support_t`` in the init call can be used to obtain the relevant information, alongside any new methods added to support type version hashing.
+
+There will be an option to opt-out of creating this service, and a way to start and stop the service after node creation as well.
+
+.. TODO::
+
+    (wjwwood) A thought just occurred to me, which is that we could maybe have "lazy" Service Servers, which watch for any Service Clients to come up on their Service name and when they see it, they could start a Service Server.
```

Comment:
What overhead is presented by a ServiceServer? Unlike a Publisher, it doesn't do work unless called, so I'm not seeing a comparison to the lazy publisher. Is it discovery overhead that's the main concern, or is there also a non-negligible memory/cpu contribution to creating the server? How does that overhead compare to the monitoring required to see whether there are any prospective Clients?